File: web/packages/protocol/src/wire.ts
/** Protobuf wire codec — must produce the same bytes as `converge-proto::wire` (see fixtures/wire.json). */
import { create, fromBinary, toBinary } from "@bufbuild/protobuf";
import type { Op, OpId, OpKind, Value, ObjectKind as EngineKind, Hlc } from "@converge/engine";
import * as pb from "./gen/converge/v1/protocol_pb.js";
import type { ByeReason, CatchUp, ClientMsg, NackReason, PresenceEntry, PresenceState, ServerMsg, UserInfo } from "./messages.js";

// ----- core types -----

function idToPb(id: OpId): pb.OpId {
  return create(pb.OpIdSchema, { replica: id.replica, counter: id.counter });
}
function idFromPb(p: pb.OpId | undefined, what: string): OpId {
  if (!p) throw new Error(`missing ${what}`);
  return { replica: BigInt.asUintN(64, p.replica), counter: p.counter };
}
function hlcToPb(h: Hlc): pb.Hlc {
  return create(pb.HlcSchema, { wallMs: h.wall, logical: h.logical });
}
function hlcFromPb(p: pb.Hlc | undefined): Hlc {
  if (!p) throw new Error("missing hlc");
  if (p.logical > 0xffff) throw new Error("logical out of range");
  return { wall: p.wallMs, logical: p.logical };
}
function valueToPb(v: Value): pb.Value {
  switch (v.t) {
    case "null": return create(pb.ValueSchema, { kind: { case: "null", value: create(pb.NullSchema) } });
    case "bool": return create(pb.ValueSchema, { kind: { case: "bool", value: v.v } });
    case "i64": return create(pb.ValueSchema, { kind: { case: "i64", value: v.v } });
    case "f64": return create(pb.ValueSchema, { kind: { case: "f64", value: v.v } });
    case "str": return create(pb.ValueSchema, { kind: { case: "str", value: v.v } });
    case "color": return create(pb.ValueSchema, { kind: { case: "color", value: v.v >>> 0 } });
    case "frac": return create(pb.ValueSchema, { kind: { case: "fracIndex", value: v.v } });
    case "ref": return create(pb.ValueSchema, { kind: { case: "objRef", value: idToPb(v.v) } });
  }
}
function valueFromPb(p: pb.Value | undefined): Value {
  const k = p?.kind;
  if (!k || k.case === undefined) throw new Error("missing value");
  switch (k.case) {
    case "null": return { t: "null" };
    case "bool": return { t: "bool", v: k.value };
    case "i64": return { t: "i64", v: k.value };
    case "f64": return { t: "f64", v: k.value };
    case "str": return { t: "str", v: k.value };
    case "color": return { t: "color", v: k.value >>> 0 };
    case "fracIndex": return { t: "frac", v: k.value };
    case "objRef": return { t: "ref", v: idFromPb(k.value, "obj_ref") };
  }
}
function entriesToPb(e: ReadonlyArray<readonly [string, Value]>): pb.Entry[] {
  return e.map(([key, value]) => create(pb.EntrySchema, { key, value: valueToPb(value) }));
}
function entriesFromPb(e: pb.Entry[]): Array<readonly [string, Value]> {
  return e.map((en) => [en.key, valueFromPb(en.value)] as const);
}
const KIND_TO_PB: Record<EngineKind, pb.ObjectKind> = { rect: pb.ObjectKind.RECT, text: pb.ObjectKind.TEXT, group: pb.ObjectKind.GROUP, connector: pb.ObjectKind.CONNECTOR, ellipse: pb.ObjectKind.ELLIPSE, line: pb.ObjectKind.LINE };
function kindFromPb(k: pb.ObjectKind): EngineKind {
  switch (k) {
    case pb.ObjectKind.RECT: return "rect";
    case pb.ObjectKind.TEXT: return "text";
    case pb.ObjectKind.GROUP: return "group";
    case pb.ObjectKind.CONNECTOR: return "connector";
    case pb.ObjectKind.ELLIPSE: return "ellipse";
    case pb.ObjectKind.LINE: return "line";
    default: throw new Error("bad object kind");
  }
}

export function opToPb(op: Op): pb.Op {
  const k = op.kind;
  let kind: pb.Op["kind"];
  switch (k.op) {
    case "create": kind = { case: "create", value: create(pb.CreateSchema, { kind: KIND_TO_PB[k.kind], props: entriesToPb(k.props) }) }; break;
    case "set_props": kind = { case: "setProps", value: create(pb.SetPropsSchema, { object: idToPb(k.object), entries: entriesToPb(k.entries) }) }; break;
    case "delete": kind = { case: "delete", value: create(pb.DeleteSchema, { object: idToPb(k.object) }) }; break;
    case "restore": kind = { case: "restore", value: create(pb.RestoreSchema, { object: idToPb(k.object) }) }; break;
  }
  return create(pb.OpSchema, { id: idToPb(op.id), hlc: hlcToPb(op.hlc), kind });
}

export function opFromPb(p: pb.Op): Op {
  const k = p.kind;
  let kind: OpKind;
  switch (k.case) {
    case "create": kind = { op: "create", kind: kindFromPb(k.value.kind), props: entriesFromPb(k.value.props) }; break;
    case "setProps": kind = { op: "set_props", object: idFromPb(k.value.object, "object"), entries: entriesFromPb(k.value.entries) }; break;
    case "delete": kind = { op: "delete", object: idFromPb(k.value.object, "object") }; break;
    case "restore": kind = { op: "restore", object: idFromPb(k.value.object, "object") }; break;
    default: throw new Error("missing op kind");
  }
  return { id: idFromPb(p.id, "op.id"), hlc: hlcFromPb(p.hlc), kind };
}

export function encodeOp(op: Op): Uint8Array {
  return toBinary(pb.OpSchema, opToPb(op));
}
export function decodeOp(bytes: Uint8Array): Op {
  return opFromPb(fromBinary(pb.OpSchema, bytes));
}

// ----- protocol types -----

function userToPb(u: UserInfo): pb.UserInfo {
  return create(pb.UserInfoSchema, { name: u.name, color: u.color >>> 0 });
}
function userFromPb(p: pb.UserInfo | undefined): UserInfo {
  return p ? { name: p.name, color: p.color >>> 0 } : { name: "", color: 0 };
}
function presenceToPb(p: PresenceState): pb.PresenceState {
  return create(pb.PresenceStateSchema, {
    hasCursor: p.cursor !== null, cursorX: p.cursor?.[0] ?? 0, cursorY: p.cursor?.[1] ?? 0, selection: p.selection.map(idToPb),
  });
}
function presenceFromPb(p: pb.PresenceState | undefined): PresenceState {
  if (!p) return { cursor: null, selection: [] };
  return { cursor: p.hasCursor ? [p.cursorX, p.cursorY] : null, selection: p.selection.map((id) => idFromPb(id, "selection")) };
}
function entryToPb(e: PresenceEntry): pb.PresenceEntry {
  return create(pb.PresenceEntrySchema, { replica: e.replica, user: userToPb(e.user), state: presenceToPb(e.state) });
}
function entryFromPb(p: pb.PresenceEntry): PresenceEntry {
  return { replica: BigInt.asUintN(64, p.replica), user: userFromPb(p.user), state: presenceFromPb(p.state) };
}

export function encodeClient(m: ClientMsg): Uint8Array {
  let msg: pb.ClientMsg["msg"];
  switch (m.type) {
    case "hello":
      msg = { case: "hello", value: create(pb.HelloSchema, {
        version: m.hello.version, doc: m.hello.doc, replica: m.hello.replica, lastSeq: m.hello.lastSeq,
        wantSnapshot: m.hello.wantSnapshot, user: userToPb(m.hello.user),
      }) };
      break;
    case "submit": msg = { case: "submit", value: create(pb.SubmitSchema, { ops: m.ops.map(opToPb) }) }; break;
    case "presence": msg = { case: "presence", value: presenceToPb(m.state) }; break;
    case "pong": msg = { case: "pong", value: create(pb.PongSchema, { nonce: m.nonce }) }; break;
  }
  return toBinary(pb.ClientMsgSchema, create(pb.ClientMsgSchema, { msg }));
}

export function decodeClient(bytes: Uint8Array): ClientMsg {
  const p = fromBinary(pb.ClientMsgSchema, bytes);
  const m = p.msg;
  switch (m.case) {
    case "hello": return { type: "hello", hello: {
      version: m.value.version, doc: m.value.doc, replica: BigInt.asUintN(64, m.value.replica), lastSeq: m.value.lastSeq,
      wantSnapshot: m.value.wantSnapshot, user: userFromPb(m.value.user),
    } };
    case "submit": return { type: "submit", ops: m.value.ops.map(opFromPb) };
    case "presence": return { type: "presence", state: presenceFromPb(m.value) };
    case "pong": return { type: "pong", nonce: m.value.nonce };
    default: throw new Error("missing client message");
  }
}

const BYE_TO_PB: Record<ByeReason, pb.ByeReason> = {
  version_mismatch: pb.ByeReason.VERSION_MISMATCH, unknown_doc: pb.ByeReason.UNKNOWN_DOC, superseded: pb.ByeReason.SUPERSEDED,
  clock_skew: pb.ByeReason.CLOCK_SKEW, slow_consumer: pb.ByeReason.SLOW_CONSUMER, restart: pb.ByeReason.RESTART, protocol_error: pb.ByeReason.PROTOCOL_ERROR,
};
const BYE_FROM_PB = Object.fromEntries(Object.entries(BYE_TO_PB).map(([k, v]) => [v, k])) as Record<number, ByeReason>;
const NACK_TO_PB: Record<NackReason, pb.NackReason> = { clock_skew: pb.NackReason.CLOCK_SKEW, malformed: pb.NackReason.MALFORMED, rejected: pb.NackReason.REJECTED };
const NACK_FROM_PB = Object.fromEntries(Object.entries(NACK_TO_PB).map(([k, v]) => [v, k])) as Record<number, NackReason>;

function catchUpToPb(c: CatchUp): pb.Welcome["catchUp"] {
  return c.kind === "ops"
    ? { case: "ops", value: create(pb.CatchUpOpsSchema, { ops: c.ops.map((e) => create(pb.CommittedOpSchema, { seq: e.seq, op: opToPb(e.op) })) }) }
    : { case: "snapshot", value: create(pb.SnapshotSchema, { bytes: c.bytes, seq: c.seq }) };
}

export function encodeServer(m: ServerMsg): Uint8Array {
  let msg: pb.ServerMsg["msg"];
  switch (m.type) {
    case "welcome":
      msg = { case: "welcome", value: create(pb.WelcomeSchema, {
        session: m.session, serverTimeMs: m.serverTimeMs, durableHeadSeq: m.durableHeadSeq, catchUp: catchUpToPb(m.catchUp), presence: m.presence.map(entryToPb),
      }) };
      break;
    case "bye": msg = { case: "bye", value: create(pb.ByeSchema, { reason: BYE_TO_PB[m.reason] }) }; break;
    case "commit": msg = { case: "commit", value: create(pb.CommitSchema, { seq: m.seq, op: opToPb(m.op) }) }; break;
    case "ack": msg = { case: "ack", value: create(pb.AckSchema, m.seq === null ? { opId: idToPb(m.opId) } : { opId: idToPb(m.opId), seq: m.seq }) }; break;
    case "nack": msg = { case: "nack", value: create(pb.NackSchema, { opId: idToPb(m.opId), reason: NACK_TO_PB[m.reason] }) }; break;
    case "presence_update": msg = { case: "presenceUpdate", value: entryToPb(m.entry) }; break;
    case "presence_leave": msg = { case: "presenceLeave", value: create(pb.PresenceLeaveSchema, { replica: m.replica }) }; break;
    case "ping": msg = { case: "ping", value: create(pb.PingSchema, { nonce: m.nonce }) }; break;
  }
  return toBinary(pb.ServerMsgSchema, create(pb.ServerMsgSchema, { msg }));
}

export function decodeServer(bytes: Uint8Array): ServerMsg {
  const m = fromBinary(pb.ServerMsgSchema, bytes).msg;
  switch (m.case) {
    case "welcome": {
      const w = m.value;
      let catchUp: CatchUp;
      if (w.catchUp.case === "ops") {
        catchUp = { kind: "ops", ops: w.catchUp.value.ops.map((c) => { if (!c.op) throw new Error("missing op"); return { seq: c.seq, op: opFromPb(c.op) }; }) };
      } else if (w.catchUp.case === "snapshot") {
        catchUp = { kind: "snapshot", bytes: w.catchUp.value.bytes, seq: w.catchUp.value.seq };
      } else {
        throw new Error("missing catch_up");
      }
      return { type: "welcome", session: w.session, serverTimeMs: w.serverTimeMs, durableHeadSeq: w.durableHeadSeq, catchUp, presence: w.presence.map(entryFromPb) };
    }
    case "bye": {
      const r = BYE_FROM_PB[m.value.reason];
      if (!r) throw new Error("bad bye reason");
      return { type: "bye", reason: r };
    }
    case "commit": {
      if (!m.value.op) throw new Error("missing op");
      return { type: "commit", seq: m.value.seq, op: opFromPb(m.value.op) };
    }
    case "ack": return { type: "ack", opId: idFromPb(m.value.opId, "op_id"), seq: m.value.seq ?? null };
    case "nack": {
      const r = NACK_FROM_PB[m.value.reason];
      if (!r) throw new Error("bad nack reason");
      return { type: "nack", opId: idFromPb(m.value.opId, "op_id"), reason: r };
    }
    case "presenceUpdate": return { type: "presence_update", entry: entryFromPb(m.value) };
    case "presenceLeave": return { type: "presence_leave", replica: BigInt.asUintN(64, m.value.replica) };
    case "ping": return { type: "ping", nonce: m.value.nonce };
    default: throw new Error("missing server message");
  }
}

/** The `Codec` used by the WebSocket transport. */
export const wireCodec = { encodeClient, decodeServer };

File: crates/converge-proto/src/wire.rs
//! Protobuf wire encoding of protocol messages (`proto/converge/v1`).

use converge_core::{Hlc, ObjectKind, Op, OpId, OpKind, ReplicaId, Value};
use prost::Message;

use crate::*;

#[allow(clippy::all, missing_docs)]
pub mod pb {
    include!(concat!(env!("OUT_DIR"), "/converge.v1.rs"));
}

#[derive(thiserror::Error, Debug)]
pub enum WireError {
    #[error("protobuf decode: {0}")]
    Decode(#[from] prost::DecodeError),
    #[error("missing field {0}")]
    Missing(&'static str),
    #[error("invalid enum value for {0}")]
    BadEnum(&'static str),
    #[error("logical clock out of range")]
    Logical,
}

type R<T> = Result<T, WireError>;

// ----- core types -----

fn id_to_pb(id: &OpId) -> pb::OpId {
    pb::OpId {
        replica: id.replica.0,
        counter: id.counter,
    }
}
fn id_from_pb(p: Option<pb::OpId>, what: &'static str) -> R<OpId> {
    let p = p.ok_or(WireError::Missing(what))?;
    Ok(OpId {
        replica: ReplicaId(p.replica),
        counter: p.counter,
    })
}
fn hlc_to_pb(h: &Hlc) -> pb::Hlc {
    pb::Hlc {
        wall_ms: h.wall_ms,
        logical: h.logical as u32,
    }
}
fn hlc_from_pb(p: Option<pb::Hlc>) -> R<Hlc> {
    let p = p.ok_or(WireError::Missing("hlc"))?;
    Ok(Hlc::new(
        p.wall_ms,
        u16::try_from(p.logical).map_err(|_| WireError::Logical)?,
    ))
}
fn value_to_pb(v: &Value) -> pb::Value {
    use pb::value::Kind;
    let kind = match v {
        Value::Null => Kind::Null(pb::Null {}),
        Value::Bool(b) => Kind::Bool(*b),
        Value::I64(i) => Kind::I64(*i),
        Value::F64(f) => Kind::F64(*f),
        Value::Str(s) => Kind::Str(s.clone()),
        Value::Color(c) => Kind::Color(*c),
        Value::FracIndex(s) => Kind::FracIndex(s.clone()),
        Value::ObjRef(id) => Kind::ObjRef(id_to_pb(id)),
    };
    pb::Value { kind: Some(kind) }
}
fn value_from_pb(p: Option<pb::Value>) -> R<Value> {
    use pb::value::Kind;
    Ok(
        match p.and_then(|v| v.kind).ok_or(WireError::Missing("value"))? {
            Kind::Null(_) => Value::Null,
            Kind::Bool(b) => Value::Bool(b),
            Kind::I64(i) => Value::I64(i),
            Kind::F64(f) => Value::F64(f),
            Kind::Str(s) => Value::Str(s),
            Kind::Color(c) => Value::Color(c),
            Kind::FracIndex(s) => Value::FracIndex(s),
            Kind::ObjRef(id) => Value::ObjRef(id_from_pb(Some(id), "obj_ref")?),
        },
    )
}
fn entries_to_pb(e: &[(String, Value)]) -> Vec<pb::Entry> {
    e.iter()
        .map(|(k, v)| pb::Entry {
            key: k.clone(),
            value: Some(value_to_pb(v)),
        })
        .collect()
}
fn entries_from_pb(e: Vec<pb::Entry>) -> R<Vec<(String, Value)>> {
    e.into_iter()
        .map(|en| Ok((en.key, value_from_pb(en.value)?)))
        .collect()
}
fn kind_to_pb(k: ObjectKind) -> i32 {
    match k {
        ObjectKind::Rect => pb::ObjectKind::Rect as i32,
        ObjectKind::Text => pb::ObjectKind::Text as i32,
        ObjectKind::Group => pb::ObjectKind::Group as i32,
        ObjectKind::Connector => pb::ObjectKind::Connector as i32,
        ObjectKind::Ellipse => pb::ObjectKind::Ellipse as i32,
        ObjectKind::Line => pb::ObjectKind::Line as i32,
    }
}
fn kind_from_pb(v: i32) -> R<ObjectKind> {
    Ok(
        match pb::ObjectKind::try_from(v).map_err(|_| WireError::BadEnum("object_kind"))? {
            pb::ObjectKind::Rect => ObjectKind::Rect,
            pb::ObjectKind::Text => ObjectKind::Text,
            pb::ObjectKind::Group => ObjectKind::Group,
            pb::ObjectKind::Connector => ObjectKind::Connector,
            pb::ObjectKind::Ellipse => ObjectKind::Ellipse,
            pb::ObjectKind::Line => ObjectKind::Line,
            pb::ObjectKind::Unspecified => return Err(WireError::BadEnum("object_kind")),
        },
    )
}

pub fn op_to_pb(op: &Op) -> pb::Op {
    use pb::op::Kind;
    let kind = match &op.kind {
        OpKind::Create { kind, props } => Kind::Create(pb::Create {
            kind: kind_to_pb(*kind),
            props: entries_to_pb(props),
        }),
        OpKind::SetProps { object, entries } => Kind::SetProps(pb::SetProps {
            object: Some(id_to_pb(object)),
            entries: entries_to_pb(entries),
        }),
        OpKind::Delete { object } => Kind::Delete(pb::Delete {
            object: Some(id_to_pb(object)),
        }),
        OpKind::Restore { object } => Kind::Restore(pb::Restore {
            object: Some(id_to_pb(object)),
        }),
    };
    pb::Op {
        id: Some(id_to_pb(&op.id)),
        hlc: Some(hlc_to_pb(&op.hlc)),
        kind: Some(kind),
    }
}

pub fn op_from_pb(p: pb::Op) -> R<Op> {
    use pb::op::Kind;
    let kind = match p.kind.ok_or(WireError::Missing("op.kind"))? {
        Kind::Create(c) => OpKind::Create {
            kind: kind_from_pb(c.kind)?,
            props: entries_from_pb(c.props)?,
        },
        Kind::SetProps(s) => OpKind::SetProps {
            object: id_from_pb(s.object, "object")?,
            entries: entries_from_pb(s.entries)?,
        },
        Kind::Delete(d) => OpKind::Delete {
            object: id_from_pb(d.object, "object")?,
        },
        Kind::Restore(r) => OpKind::Restore {
            object: id_from_pb(r.object, "object")?,
        },
    };
    Ok(Op {
        id: id_from_pb(p.id, "op.id")?,
        hlc: hlc_from_pb(p.hlc)?,
        kind,
    })
}

pub fn encode_op(op: &Op) -> Vec<u8> {
    op_to_pb(op).encode_to_vec()
}
pub fn decode_op(bytes: &[u8]) -> R<Op> {
    op_from_pb(pb::Op::decode(bytes)?)
}

// ----- protocol types -----

fn user_to_pb(u: &UserInfo) -> pb::UserInfo {
    pb::UserInfo {
        name: u.name.clone(),
        color: u.color,
    }
}
fn user_from_pb(p: Option<pb::UserInfo>) -> UserInfo {
    p.map(|u| UserInfo {
        name: u.name,
        color: u.color,
    })
    .unwrap_or_default()
}
fn presence_to_pb(p: &PresenceState) -> pb::PresenceState {
    pb::PresenceState {
        has_cursor: p.cursor.is_some(),
        cursor_x: p.cursor.map(|c| c.0).unwrap_or(0.0),
        cursor_y: p.cursor.map(|c| c.1).unwrap_or(0.0),
        selection: p.selection.iter().map(id_to_pb).collect(),
    }
}
fn presence_from_pb(p: pb::PresenceState) -> R<PresenceState> {
    Ok(PresenceState {
        cursor: if p.has_cursor {
            Some((p.cursor_x, p.cursor_y))
        } else {
            None
        },
        selection: p
            .selection
            .into_iter()
            .map(|id| id_from_pb(Some(id), "selection"))
            .collect::<R<_>>()?,
    })
}
fn entry_to_pb(e: &PresenceEntry) -> pb::PresenceEntry {
    pb::PresenceEntry {
        replica: e.replica.0,
        user: Some(user_to_pb(&e.user)),
        state: Some(presence_to_pb(&e.state)),
    }
}
fn entry_from_pb(p: pb::PresenceEntry) -> R<PresenceEntry> {
    Ok(PresenceEntry {
        replica: ReplicaId(p.replica),
        user: user_from_pb(p.user),
        state: presence_from_pb(p.state.unwrap_or_default())?,
    })
}

pub fn client_msg_to_pb(m: &ClientMsg) -> pb::ClientMsg {
    use pb::client_msg::Msg;
    let msg = match m {
        ClientMsg::Hello(h) => Msg::Hello(pb::Hello {
            version: h.version,
            doc: h.doc.0.clone(),
            replica: h.replica.0,
            last_seq: h.last_seq,
            want_snapshot: h.want_snapshot,
            user: Some(user_to_pb(&h.user)),
        }),
        ClientMsg::Submit { ops } => Msg::Submit(pb::Submit {
            ops: ops.iter().map(op_to_pb).collect(),
        }),
        ClientMsg::Presence(p) => Msg::Presence(presence_to_pb(p)),
        ClientMsg::Pong { nonce } => Msg::Pong(pb::Pong { nonce: *nonce }),
    };
    pb::ClientMsg { msg: Some(msg) }
}

pub fn client_msg_from_pb(p: pb::ClientMsg) -> R<ClientMsg> {
    use pb::client_msg::Msg;
    Ok(match p.msg.ok_or(WireError::Missing("client_msg"))? {
        Msg::Hello(h) => ClientMsg::Hello(Hello {
            version: h.version,
            doc: DocId(h.doc),
            replica: ReplicaId(h.replica),
            last_seq: h.last_seq,
            want_snapshot: h.want_snapshot,
            user: user_from_pb(h.user),
        }),
        Msg::Submit(s) => ClientMsg::Submit {
            ops: s.ops.into_iter().map(op_from_pb).collect::<R<_>>()?,
        },
        Msg::Presence(p) => ClientMsg::Presence(presence_from_pb(p)?),
        Msg::Pong(p) => ClientMsg::Pong { nonce: p.nonce },
    })
}

fn bye_to_pb(r: ByeReason) -> i32 {
    (match r {
        ByeReason::VersionMismatch => pb::ByeReason::VersionMismatch,
        ByeReason::UnknownDoc => pb::ByeReason::UnknownDoc,
        ByeReason::Superseded => pb::ByeReason::Superseded,
        ByeReason::ClockSkew => pb::ByeReason::ClockSkew,
        ByeReason::SlowConsumer => pb::ByeReason::SlowConsumer,
        ByeReason::Restart => pb::ByeReason::Restart,
        ByeReason::ProtocolError => pb::ByeReason::ProtocolError,
    }) as i32
}
fn bye_from_pb(v: i32) -> R<ByeReason> {
    Ok(
        match pb::ByeReason::try_from(v).map_err(|_| WireError::BadEnum("bye_reason"))? {
            pb::ByeReason::VersionMismatch => ByeReason::VersionMismatch,
            pb::ByeReason::UnknownDoc => ByeReason::UnknownDoc,
            pb::ByeReason::Superseded => ByeReason::Superseded,
            pb::ByeReason::ClockSkew => ByeReason::ClockSkew,
            pb::ByeReason::SlowConsumer => ByeReason::SlowConsumer,
            pb::ByeReason::Restart => ByeReason::Restart,
            pb::ByeReason::ProtocolError => ByeReason::ProtocolError,
            pb::ByeReason::Unspecified => return Err(WireError::BadEnum("bye_reason")),
        },
    )
}
fn nack_to_pb(r: NackReason) -> i32 {
    (match r {
        NackReason::ClockSkew => pb::NackReason::ClockSkew,
        NackReason::Malformed => pb::NackReason::Malformed,
        NackReason::Rejected => pb::NackReason::Rejected,
    }) as i32
}
fn nack_from_pb(v: i32) -> R<NackReason> {
    Ok(
        match pb::NackReason::try_from(v).map_err(|_| WireError::BadEnum("nack_reason"))? {
            pb::NackReason::ClockSkew => NackReason::ClockSkew,
            pb::NackReason::Malformed => NackReason::Malformed,
            pb::NackReason::Rejected => NackReason::Rejected,
            pb::NackReason::Unspecified => return Err(WireError::BadEnum("nack_reason")),
        },
    )
}

pub fn server_msg_to_pb(m: &ServerMsg) -> pb::ServerMsg {
    use pb::server_msg::Msg;
    let msg = match m {
        ServerMsg::Welcome {
            session,
            server_time_ms,
            durable_head_seq,
            catch_up,
            presence,
        } => Msg::Welcome(pb::Welcome {
            session: session.0,
            server_time_ms: *server_time_ms,
            durable_head_seq: *durable_head_seq,
            catch_up: Some(match catch_up {
                CatchUp::Ops(ops) => pb::welcome::CatchUp::Ops(pb::CatchUpOps {
                    ops: ops
                        .iter()
                        .map(|(seq, op)| pb::CommittedOp {
                            seq: *seq,
                            op: Some(op_to_pb(op)),
                        })
                        .collect(),
                }),
                CatchUp::Snapshot { bytes, seq } => pb::welcome::CatchUp::Snapshot(pb::Snapshot {
                    bytes: bytes.clone(),
                    seq: *seq,
                }),
            }),
            presence: presence.iter().map(entry_to_pb).collect(),
        }),
        ServerMsg::Bye { reason } => Msg::Bye(pb::Bye {
            reason: bye_to_pb(*reason),
        }),
        ServerMsg::Commit { seq, op } => Msg::Commit(pb::Commit {
            seq: *seq,
            op: Some(op_to_pb(op)),
        }),
        ServerMsg::Ack { op_id, seq } => Msg::Ack(pb::Ack {
            op_id: Some(id_to_pb(op_id)),
            seq: *seq,
        }),
        ServerMsg::Nack { op_id, reason } => Msg::Nack(pb::Nack {
            op_id: Some(id_to_pb(op_id)),
            reason: nack_to_pb(*reason),
        }),
        ServerMsg::PresenceUpdate(e) => Msg::PresenceUpdate(entry_to_pb(e)),
        ServerMsg::PresenceLeave { replica } => {
            Msg::PresenceLeave(pb::PresenceLeave { replica: replica.0 })
        }
        ServerMsg::Ping { nonce } => Msg::Ping(pb::Ping { nonce: *nonce }),
    };
    pb::ServerMsg { msg: Some(msg) }
}

pub fn server_msg_from_pb(p: pb::ServerMsg) -> R<ServerMsg> {
    use pb::server_msg::Msg;
    Ok(match p.msg.ok_or(WireError::Missing("server_msg"))? {
        Msg::Welcome(w) => ServerMsg::Welcome {
            session: SessionId(w.session),
            server_time_ms: w.server_time_ms,
            durable_head_seq: w.durable_head_seq,
            catch_up: match w.catch_up.ok_or(WireError::Missing("catch_up"))? {
                pb::welcome::CatchUp::Ops(o) => CatchUp::Ops(
                    o.ops
                        .into_iter()
                        .map(|c| Ok((c.seq, op_from_pb(c.op.ok_or(WireError::Missing("op"))?)?)))
                        .collect::<R<_>>()?,
                ),
                pb::welcome::CatchUp::Snapshot(s) => CatchUp::Snapshot {
                    bytes: s.bytes,
                    seq: s.seq,
                },
            },
            presence: w
                .presence
                .into_iter()
                .map(entry_from_pb)
                .collect::<R<_>>()?,
        },
        Msg::Bye(b) => ServerMsg::Bye {
            reason: bye_from_pb(b.reason)?,
        },
        Msg::Commit(c) => ServerMsg::Commit {
            seq: c.seq,
            op: op_from_pb(c.op.ok_or(WireError::Missing("op"))?)?,
        },
        Msg::Ack(a) => ServerMsg::Ack {
            op_id: id_from_pb(a.op_id, "op_id")?,
            seq: a.seq,
        },
        Msg::Nack(n) => ServerMsg::Nack {
            op_id: id_from_pb(n.op_id, "op_id")?,
            reason: nack_from_pb(n.reason)?,
        },
        Msg::PresenceUpdate(e) => ServerMsg::PresenceUpdate(entry_from_pb(e)?),
        Msg::PresenceLeave(l) => ServerMsg::PresenceLeave {
            replica: ReplicaId(l.replica),
        },
        Msg::Ping(p) => ServerMsg::Ping { nonce: p.nonce },
    })
}

pub fn encode_client(m: &ClientMsg) -> Vec<u8> {
    client_msg_to_pb(m).encode_to_vec()
}
pub fn decode_client(bytes: &[u8]) -> R<ClientMsg> {
    client_msg_from_pb(pb::ClientMsg::decode(bytes)?)
}
pub fn encode_server(m: &ServerMsg) -> Vec<u8> {
    server_msg_to_pb(m).encode_to_vec()
}
pub fn decode_server(bytes: &[u8]) -> R<ServerMsg> {
    server_msg_from_pb(pb::ServerMsg::decode(bytes)?)
}
